### PR TITLE
align selections

### PR DESCRIPTION
**Pull request by QiBaobin**
Screen Shot 2021-11-14 at 10 01 06 PM

Select above lines, `&` and input `\|` would turn them to 

Screen Shot 2021-11-14 at 10 01 25 PM

`2&` and input `\|` would turn them to 
Screen Shot 2021-11-14 at 10 01 38 PM

`3&` and input `\|` would turn them to 
Screen Shot 2021-11-14 at 10 01 52 PM

### CODE REVIEW

File: helix-term/src/keymap.rs [hunk 1]
```diff
@@ -601,6 +601,7 @@ impl Default for Keymaps {
             // "q" => record_macro,
             // "Q" => replay_macro,
 
+            "&" => align_lines,
             // & align selections
```
Comment (Omnikar) [inline helix-term/src/keymap.rs:605]:
```suggestion
```

**Comment by archseer** [comment]
Rather than using a regex for the delimiter, this should align selections in the same line: http://kakoune.org/video/align.webm

^ it selects the whole line, splits into sub selections, moves cursors to the start of the selection, then `&` aligns the cursors.

https://github.com/mawww/kakoune/blob/e7100dc87434d933fd8268e5bf70080b627750c5/src/normal.cc#L1615-L1701

**Comment by QiBaobin** [comment]
@archseer , I tried to implement using selections, one problem is that we the cursor is not at the selection we want(in another word, it's at a part of the separators), it would make cursor as a selection too.

**Comment by archseer** [comment]
What do you mean? Can you show me an example?

 In general we want commands to always take advantage of multiple selections, it's more idiomatic for helix.

**Comment by QiBaobin** [comment]
When I used split_selections with `\|` after selecting two lines, I got below selections, as you can see the selections in the first line doesn't make sense.

Screen Shot 2021-11-15 at 9 58 24 AM


Just checked select_regex, it seems to work in this case.

**Comment by archseer** [comment]
Here's something that works for me: select all the columns then use `s\w<enter>`, `mi|`, `_` and then `alt-;` to jump to the start.

 https://asciinema.org/a/q7Jd8EgPQ6OVMZ4aAWCTGTjJt

**Comment by QiBaobin** [comment]
Updated the code, shall work well with select_regex, but maybe not with split_selection without move cursor.

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -625,6 +626,64 @@ fn trim_selections(cx: &mut Context) {
...
+    doc.append_changes_to_history(view.id);
+}
+
+fn align_fragment_to_width(fragment: &str, width: usize, align_style: usize) -> String {
+    let trimed = fragment.trim_matches(|c| c == ' ');
+    let mut s = " ".repeat(width - trimed.chars().count());
+    match align_style {
+        2 => s.insert_str(s.len() / 2, trimed), // center align
+        3 => s.push_str(trimed),                // right align
+        _ => s.insert_str(0, trimed),           // left align
```
Comment (pickfire) [inline helix-term/src/commands.rs:682]: Shouldn't this be 1? Then we error on the rest?

Comment (QiBaobin) [reply helix-term/src/commands.rs:688]: Added a check to only accept 1, 2, 3 as user count.

Comment (pickfire) [reply helix-term/src/commands.rs:688]:
```suggestion
        1 => s.insert_str(0, trimed),           // left align
        n => unimplemented!(n),
```
I think we should do it this way, you may want to realign.

File: helix-term/src/commands.rs [hunk 3]
```diff
@@ -625,6 +626,71 @@ fn trim_selections(cx: &mut Context) {
...
 }
 
+// align text in selection
+fn align_selections(cx: &mut Context) {
+    let align_style = cx.count();
+    if align_style > 3 {
+        cx.editor.set_error(
+            "align only accept 1,2,3 as count to set left/center/right align".to_string(),
+        );
+    }
```
Comment (sudormrfbin) [inline helix-term/src/commands.rs:636]: There should be a `return` here else it panics because of `unimplemented!` further down.

Comment (QiBaobin) [reply helix-term/src/commands.rs:637]: Good catch, thanks!

File: helix-term/src/commands.rs [hunk 4]
```diff
@@ -625,6 +626,72 @@ fn trim_selections(cx: &mut Context) {
...
+    doc.append_changes_to_history(view.id);
+}
+
+fn align_fragment_to_width(fragment: &str, width: usize, align_style: usize) -> String {
+    let trimed = fragment.trim_matches(|c| c == ' ');
+    let mut s = " ".repeat(width - trimed.chars().count());
+    match align_style {
+        2 => s.insert_str(s.len() / 2, trimed), // center align
+        3 => s.push_str(trimed),                // right align
+        1 => s.insert_str(0, trimed),           // left align
```
Comment (pickfire) [inline helix-term/src/commands.rs:689]:
```suggestion
        1 => s.insert_str(0, trimed),           // left align
        2 => s.insert_str(s.len() / 2, trimed), // center align
        3 => s.push_str(trimed),                // right align
```

File: helix-term/src/commands.rs [hunk 5]
```diff
@@ -625,6 +626,72 @@ fn trim_selections(cx: &mut Context) {
...
+            }
+        } else {
+            column_widths.push(sel.to() - sel.from());
+        }
+    }
+    last_line = text.len_lines();
+    let transaction = Transaction::change_by_selection(doc.text(), selection, |range| {
+        let l = range.cursor_line(text);
+        column = if l != last_line { 0 } else { column + 1 };
+        last_line = l;
```
Comment (pickfire) [inline helix-term/src/commands.rs:667]: Can you please drop some comments here to ease understanding for future readers?

Comment (QiBaobin) [reply helix-term/src/commands.rs]: done

File: helix-term/src/commands.rs (hunk 4 above)
Comment (QiBaobin) [reply helix-term/src/commands.rs]: `swap_with_slice` need two `mut str` or string, but the trimmed or the fragment is not own by us, to use it we have make another string, which is same thing, right?

**Review by archseer — APPROVED** [review]
Great work! Tested locally 🎉

**Comment by sudormrfbin** [comment]
We also need some docs about this in the book in the usage section, especially since the workflow might be confusing at first (using cursors to align) and left, right, center alignment are controlled by counts before `&`. Bonus points for a gif :)

Awesome work btw @QiBaobin 🎉

**Pull request merged**